plugins/pixart-tp/fu-pxi-tp-firmware.c

```diff
@@ -0,0 +1,940 @@
+		return FALSE;
+
+	/* header length (LE16) */
+	if (!fu_memread_uint16_safe(d, hdrsz, PXI_TP_O_HDRLEN, &hdrlen, G_LITTLE_ENDIAN, NULL))
+		return FALSE;
+
+	if (hdrlen != PXI_TP_HEADER_V1_LEN)
+		return FALSE;
+
+	return TRUE;
+}
+
+/**
+ * fu_pxi_tp_firmware_parse:
+ * @firmware: the firmware container object
+ * @stream: unused (fwupd vfunc signature)
+ * @flags: parse flags
+ * @error: return location for a #GError, or %NULL
+ *
+ * Parses the FWHD v1.0 header, verifies header CRC32 and payload CRC32,
+ * and decodes up to 8 section descriptors into a private array. For
+ * internal sections, we trust `internal_file_start` as an absolute file
+ * offset from the beginning of the blob (already including header).
+ *
+ * Returns: %TRUE on success, %FALSE with @error set on failure.
+ */
+/* nocheck:memread */
+static gboolean
+fu_pxi_tp_firmware_parse(FuFirmware *firmware,
+			 GInputStream *stream,
+			 FuFirmwareParseFlags flags,
+			 GError **error)
+{
+	const guint8 *d = NULL;
+	gsize sz = 0;
+	guint64 sectab_end = 0;
+	FuPxiTpFirmware *self = FU_PXI_TP_FIRMWARE(firmware);
+
+	/* read the entire blob through the fwupd streaming helper */
+	g_autoptr(GBytes) fw = fu_input_stream_read_bytes(stream, 0, G_MAXSIZE, NULL, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* store original bytes for later (patching/export/slices) */
+	fu_firmware_set_bytes(firmware, fw);
+
+	d = g_bytes_get_data(fw, NULL);
+	sz = g_bytes_get_size(fw);
+
+	/* lightweight checks again (with errors reported here) */
+	if (sz < PXI_TP_HEADER_V1_LEN) {
+		g_debug("file too small for header: %" G_GSIZE_FORMAT " bytes (need >= %u)",
+			sz,
+			(guint)PXI_TP_HEADER_V1_LEN);
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "file too small for FWHD header");
+		return FALSE;
+	}
+	if (memcmp(d + PXI_TP_O_MAGIC, PXI_TP_MAGIC, 4) != 0) {
+		g_debug("bad magic, expected 'FWHD'");
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "invalid FWHD magic");
+		return FALSE;
+	}
+
+	/* --- parse fixed header (all via safe readers) --- */
+	if (!fu_memread_uint16_safe(d,
+				    sz,
+				    PXI_TP_O_HDRLEN,
+				    &self->header_len,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+	if (self->header_len != PXI_TP_HEADER_V1_LEN || self->header_len > sz ||
+	    self->header_len < 4) {
+		g_debug("header length error: header_len=%u, file_size=%" G_GSIZE_FORMAT,
+			(guint)self->header_len,
+			sz);
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "invalid FWHD header length");
+		return FALSE;
+	}
+
+	if (!fu_memread_uint16_safe(d,
+				    sz,
+				    PXI_TP_O_HDRVER,
+				    &self->header_ver,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+	if (!fu_memread_uint16_safe(d,
+				    sz,
+				    PXI_TP_O_FILEVER,
+				    &self->file_ver,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
```

Comment:
I think using `#[derive(ParseStream)]` and defining `FuPxiTpFirmwareHdr` in the .rs file would make this much nicer indeed!